tuftool/src/clone.rs

```diff
@@ -0,0 +1,144 @@
+        let root_path = if let Some(path) = &self.root {
+            PathBuf::from(path)
+        } else if self.allow_root_download {
+            let outdir = std::env::current_dir().context(error::CurrentDir)?;
+            download_root(&self.metadata_base_url, self.root_version, outdir)?
+        } else {
+            eprintln!("No root.json available");
+            std::process::exit(1);
+        };
+
+        // Structopt won't allow `targets_base_url` to be None when it is required.  We require the
+        // user to supply `targets_base_url` in the case they actually plan to download targets.
+        // When downloading metadata, we don't ever need to access the targets URL, so we use a
+        // fake URL to satisfy the library.
+        let targets_base_url = self
+            .targets_base_url
+            .as_ref()
+            .unwrap_or(&Url::parse(UNUSED_URL).context(error::UrlParse {
+                url: UNUSED_URL.to_owned(),
+            })?)
+            .clone();
+
+        // Load repository
+        let expiration_enforcement = if self.allow_expired_repo {
+            expired_repo_warning(&self.metadata_dir);
+            ExpirationEnforcement::Unsafe
+        } else {
+            ExpirationEnforcement::Safe
+        };
+        let repository = RepositoryLoader::new(
+            File::open(&root_path).context(error::OpenRoot { path: &root_path })?,
+            self.metadata_base_url.clone(),
+            targets_base_url,
+        )
+        .expiration_enforcement(expiration_enforcement)
+        .load()
+        .context(error::RepoLoad)?;
+
+        // Clone the repository, downloading none, all, or a subset of targets
+        let clone_result = if self.metadata_only {
+            println!("Cloning repository metadata to {:?}", self.metadata_dir);
+            repository.cache_metadata(&self.metadata_dir, true)
+        } else {
+            // Similar to `targets_base_url, structopt's guard rails won't let us have a
+            // `targets_dir` that is None when the argument is required.  We only require the user
+            // to supply a targets directory if they actually plan on downloading targets.
+            let targets_dir = self.targets_dir.as_ref().expect(
+                "Developer error: `targets_dir` is required unless downloading metadata only",
+            );
+
+            if self.target_names.is_empty() {
+                println!(
+                    "Cloning repository:\n\tmetadata location: {:?}\n\ttargets location: {:?}",
+                    self.metadata_dir, targets_dir
+                );
+                // The None turbofish is required to satisfy the compiler, sorry.
+                repository.cache(&self.metadata_dir, &targets_dir, None::<&[&str]>, true)
+            } else {
+                println!(
+                    "Cloning repository with a subset of targets:\n\tmetadata location:{:?}\n\ttargets location: {:?}",
```

Comment:
nit: it doesn't feel like this slightly different message is super valuable; the user knows they requested a subset.  Should we just always pass the list of targets?  Could remove the Option to solve #405..?

Reply:
Oh gosh, I was being silly.  I'll always pass the list of targets and remove the nested `if` altogether.  #405 can happen separately.

Reply:
Oops - I can't actually do that.  `None` and an empty list are treated differently so I'll need to keep the extra `if` statement for now.  I'd rather not  fix #405 as a part of this PR, especially because that "fix" changes the API of `cache()` which effectively breaks anyone who is using the library.

All that is to say I'll change my code to a single print message, but keep the `if` statement.